### DISCUSSION TITLE
Print PDF reports with Rhino GH and C#

### DISCUSSION THREAD

**Discussion by kenchluu** [idea-statica/ideastatica-public#731] [Rhino/Grasshopper] [answered]
I have tried the print PDF report function in C# and GH, but both give me the same following errors. Please advise how to resolve

System.ServiceModel.FaultException: 'The server was unable to process the request due to an internal error.  For more information about the error, either turn on IncludeExceptionDetailInFaults (either from ServiceBehaviorAttribute or from the <serviceDebug> configuration behavior) on the server in order to send the exception 

![Error export PDF_CSharp](https://github.com/user-attachments/assets/0f664c27-58b0-4300-836a-a73a3a6f26c7)
![Error export PDF_GH](https://github.com/user-attachments/assets/e09a90d0-22d5-4b89-8647-73ab50adb794)

**Comment by nathanlukeIDEA** [comment]
Hi @kenchluu, can you please tell me what version of IDEA StatiCa you are using?

**Comment by kenchluu** [comment]
I'm using IDEA Statica Version 24.1.0.3009

**Comment by nathanlukeIDEA** [comment]
Hi Ken, I am getting the same behaviour with ConHiddenCalculator. I would suggest trying to update your App to our new Connection API. The report output should work here for your version. https://github.com/idea-statica/ideastatica-public/blob/main/src/api-sdks/connection-api/clients/csharp/examples/CodeSamples/Samples/GeneratePdf.cs

We are in the process now of updating GH to use this new API.

**Comment by kenchluu** [comment]
Hi Nathan,

Thanks for your prompt reply. About the code, how can I declare the ConnectionApiClient conClient

I have tried sth like this but it does not work:  ConnectionApiClient conClient = new ConnectionApiClient("http://localhost:5000");

Not sure what value we should use for the Basepath

![image](https://github.com/user-attachments/assets/40d7057b-594f-4d88-92fd-c59a66650652)

**Reply by kenchluu** [reply]
Hi Nathan, can you please advise my query?

**Reply by nathanlukeIDEA** [reply]
Hi @kenchluu,

**NOTE: we are making an update to the below client creation code in a day or two to enable automatic starting of the service.**

But to respond...

Here is the program code that creates the client. 
https://github.com/idea-statica/ideastatica-public/blob/main/src/api-sdks/connection-api/clients/csharp/examples/CodeSamples/Program.cs

Simplified version of the above is below. You can use this code:


```csharp
// Create the client which is connected to the service.
ConnectionApiClientFactory clientFactory = new ConnectionApiClientFactory("http://localhost:5000");
using (var conClient = await clientFactory.CreateConnectionApiClient())
{

//PDF Creation Code

}
```



You will need to start the service to be able to run the API atm. 

We currently only support connecting to a service running on a localhost (eg. 'http://localhost:5000/'). To start the service, manually navigate to the "C:\Program Files\IDEA StatiCa\StatiCa 24.1\net6.0-windows" folder. Using CLI:

```cli
IdeaStatiCa.ConnectionRestApi.exe 
```

**Reply by kenchluu** [reply]
Hi Nathan,

I have followed the instruction but still got this error

![image](https://github.com/user-attachments/assets/4c2e41d3-4d9b-4a14-8ac6-acb197198341)


By the way the code got error when I declare "conClient" as ConnectionApiClient
After I change the type to IConnectionApiClient, there is no more error however not sure if it is correct

![image](https://github.com/user-attachments/assets/7365f1f3-6435-4939-af32-a4a88b8bcd8a)

**Reply by nathanlukeIDEA** [reply, marked as answer]
Hi @kenchluu, IConnectionApiClient is correct. In my examples, the casting to ConnectionApiClient is done by the invoke method but I will update them to IConnectionApiClient. Let me know if any other issues.

**Comment by kenchluu** [comment]
Hi Nathan,

You will update in the next version of IdeaStatiCa.ConnectionApi right? Please let me know once it is updated. Thank you

**Comment by nathanlukeIDEA** [comment]
Hi @kenchluu, this should be done in the latest examples now. Cheers Nathan

**Reply by kenchluu** [reply]
Hi @nathanlukeIDEA, thanks for the update, I tried the latest example and got the following error. Can you please support? Thanks

![image](https://github.com/user-attachments/assets/0b5e7129-e2e1-44b6-96d7-6de286840f71)

![image](https://github.com/user-attachments/assets/d018502e-e734-4101-92cd-f1685478e767)

**Reply by nathanlukeIDEA** [reply]
Hi @kenchluu, I think it is because _net6.0-windows_ in at the end of the runner path. If you remove this and see if it works please.

**Reply by kenchluu** [reply]
Hi @nathanlukeIDEA ,

The use net6.0-windows because the IdeaStatiCa.ConnectionRestApi.exe is located in the net6.0-windows folder. If I don't use it, an error occurs.

![image](https://github.com/user-attachments/assets/844febd2-d423-485a-a82e-1ff35db5b96e)

**Comment by kenchluu** [comment]
Hi @nathanlukeIDEA ,

I tried to run IdeaStatiCa.ConnectionRestApi.exe separately then select 1 -> Attach to an ready running API service, it no longer have that error, however it still have error at the SaveReportPDFAsync function.

![image](https://github.com/user-attachments/assets/6c4f0ae4-1bbd-49d8-b163-e37d371a41a2)


I thought there is issue with the API Service however when I run  all other examples it works except Example 8 & 9, so it seems that the API Service works ok, I dont know why it is still error at this SaveReportPDFAsync function.

**Reply by nathanlukeIDEA** [reply]
Hi @kenchluu, we are still working to fix the PDF generation function. It is a known issue. I will get back to you.

**Comment by kenchluu** [comment]
Hi @nathanlukeIDEA , thank you, I'm looking forward to the updates.
I’ve noticed that whenever we export a report, the IdeaStatica window has to open each time, and this process takes quite a while. Is there a way to run it in the background or speed it up somehow?
If we need to export a large number of reports, the current method feels a bit inconvenient. Just a quick comment!

**Reply by nathanlukeIDEA** [reply]
Hi, @kenchluu, we are looking at improving the Report generation through the API. Stay tuned for this. Cheers

**Reply by kenchluu** [reply]
Thanks @nathanlukeIDEA

**Comment by nathanlukeIDEA** [comment]
Hi @kenchluu, In Version 25.1 report functionality has been improved, and the App no longer needs to open. We also support report creation accross multiple connections in the connection project. I hope it is a big improvement for you.

**Comment by kenchluu** [comment]
Hi Nathan,

Thanks for the update. 

The new report-exporting function is great. It is a really a big improvement that I expect. Although, the C# code is still giving me errors, so I’ve switched to Python instead, and it works perfectly.

I have one more question: is there a way to retrieve the updated connection image after modifying the parameters? If so, could you please share an example using Python?

Thanks.

**Reply by nathanlukeIDEA** [reply]
Hi @kenchluu, have you tested this example? https://github.com/idea-statica/ideastatica-public/blob/main/src/api-sdks/connection-api/clients/csharp/examples/CodeSamples/Samples/GeneratePdf.cs It is working for me.
